lvsfunc/comparison.py

```diff
@@ -13,6 +17,316 @@
+        for clip, name in self.clips:
+            try:
+                self.formats.append(clip.format.name)
+            except AttributeError:
+                self.formats.append(clip.format)
+        self.format: Optional[str] = self.formats[0] if len(set(self.formats)) == 1 else None
+
+    def _smart_crop(self, direction: Direction = Direction.HORIZONTAL):
+        if not self.width or not self.height:
+            raise ValueError("Comparer: all clips must have same width and height if cropping")
+        breaks_subsampling = ((direction == Direction.HORIZONTAL and (((self.width // self.num_clips) % 2)
+                                                                      or ((self.width % self.num_clips) % 2)))
+                              or (direction == Direction.VERTICAL and (((self.height // self.num_clips) % 2)
+                                                                       or ((self.height % self.num_clips) % 2))))
+
+        if breaks_subsampling and not all(get_subsampling(clip) in ('444', None) for clip in self.clips.values()):
+            raise ValueError("Comparer: resulting cropped width or height violates subsampling rules; "
+                             "consider resampling to YUV444 or RGB before attempting to crop")
+
+        if direction == Direction.HORIZONTAL:
+            crop_width, overflow = self.width // self.num_clips, self.width % self.num_clips
+
+            for key, (name, clip) in enumerate(self.clips.items()):
+                left_crop = crop_width * key
+                right_crop = crop_width * (self.num_clips - 1 - key)
+
+                if key != (self.num_clips - 1):
+                    right_crop += overflow
+
+                self.clips[name] = clip.std.Crop(left=left_crop, right=right_crop)
+
+        elif direction == Direction.VERTICAL:
+            crop_height, overflow = self.height // self.num_clips, self.height % self.num_clips
+
+            for key, (name, clip) in enumerate(self.clips.items()):
+                top_crop = crop_height * key
+                bottom_crop = crop_height * (self.num_clips - 1 - key)
+
+                if key != (self.num_clips - 1):
+                    bottom_crop += overflow
+
+                self.clips[name] = clip.std.Crop(top=top_crop, bottom=bottom_crop)
+
+    def _markclips(self) -> List[vs.VideoNode]:
+        """
+        This allows for internal functions to specify a space-only string,
+        such as '   ' that avoid placing a label on the clip.
+
+        By default, `vapoursynth.core.text.Text(text='')` will not place a label,
+        but specifying multiple clips in the **clips Dict with the same '' key will not work.
+        """
+        clips: List[vs.VideoNode] = []
+        for name, clip in self.clips.items():
+            if name.strip():
+                clips.append(clip.text.Text(text=name, alignment=self.label_alignment))
+            else:
+                clips.append(clip)
+
+        return clips
```

Comment:
```suggestion
        return [clip.text.Text(text=name, alignment=self.label_alignment)
                if name.strip()
                else clip
                for name, clip in self.clips.items()]
```

Comprehensions are nice.

Are VideoNodes hashable? If yes, return a set to reflect unorderedness.